### ISSUE TITLE
Cleanup/purge users and contractors that no longer work on Backdrop or are spam

### ISSUE THREAD

**Issue by bugfolder** [backdrop-ops/backdropcms.org#999]
Follow-up to [this Zulip conversation](https://backdrop.zulipchat.com/#narrow/stream/352942-Marketing/topic/Landing.20Pages):

> we really neeed to do a purge of the services and contractors that no longer work on Backdrop or are spam.

The question, then, is what process do we implement to ensure that data for contractors and services remains current and up to date? Opening this issue to start the discussion.

Currently there are 76 contractors listed on [Contractors for Hire](https://backdropcms.org/support/services/contractors?expertise=All&country=All). It's a View of user accounts that have selected "Available for hire" in their profile.

On [Backdrop services](https://backdropcms.org/support/services), under Professional Services there are currently 58 organizations listed. These are nodes of type "Organization profile".

I will mention one approach, which we use on our site that lists local origami groups; the group listings are nodes that can be created by users. Since such groups come and go, we have the same problem: we need to keep them up to date. So after a year has passed since the node was last updated, we start sending periodic automated emails to the creator account, asking them to log in, check their data, and then re-save their listing (which then resets the "last-updated" clock). If they don't do that, they keep getting nagging emails, and then in the admin listing of all groups, we can see who's ignored us and decide if the listing should be deleted or they should receive a personal follow-up. Notably, there is no automatic deletion; it's always a human making that decision. The number of contractors and services above is not too big as to make something like that infeasible.

I'm sure there's other ways of addressing the issue, so let's hear them here!

**Comment by yorkshire-pudding** [comment]
Note also #977 
We have a number who claim they do things for the backdrop community that they do not.  Again this requires a human review (I am happy to do).  Perhaps if we could also have one or more admin views where we get to see all the contractor nodes in one place (all the org nodes in another) with all the key content fields; it's unwieldy, but still more effective than having to click through each.

I like the idea of a periodic review to ensure still accurate.  I think in this case, though the stakes are higher, in that we are directing possible customers towards them and diverting attention from genuine Backdrop contractors and service providers, so we should possibly consider unpublish if they don't update within a certain time frame.

**Comment by oadaeh** [comment]
You could automate more without major repercussions, something like:
* Send 3 nag emails in some periodic time frame (like 1 week)
* On the 4th check, if no change has been made, take some action like unpublish the node or block the user or uncheck the checkbox
* After the above action(s) is(are) taken, send a notice (maybe only weekly?) to an admin or admin group about the latest actions taken

Then things can happen without requiring a lot of human intervention until it's necessary, and then there could still be some measure of undoing things without any problem.

**Comment by yorkshire-pudding** [comment]
> On the 4th check, if no change has been made, take some action like unpublish the node or block the user or uncheck the checkbox

I think "uncheck the checkbox" fulfils the objective of removing from the contractor list without causing other issues.  The contractor for hire pages are actually user profiles so to unpublish we'd have to block.  Organisation (service provider) nodes could be unpublished.

**Comment by irinaz** [comment]
> I think "uncheck the checkbox" fulfils the objective of removing from the contractor list without causing other issues. The contractor for hire pages are actually user profiles so to unpublish we'd have to block. 

I believe that it is possible to add simple checkbox on a user profile that will

- ask to confirm that they are active for one year
- use in a view for showing/hiding in service listing

**Comment by irinaz** [comment]
@yorkshire-pudding , what needs to happen to begin work on this issue?  Do we need to document proposed solution/steps?

**Comment by yorkshire-pudding** [comment]
@jenlampton - do you have any thoughts on this?

**Comment by jenlampton** [comment]
> So after a year has passed since the node was last updated, we start sending periodic automated emails to the creator account, asking them to log in, check their data, and then re-save their listing (which then resets the "last-updated" clock). If they don't do that, they keep getting nagging emails, and then in the admin listing of all groups, we can see who's ignored us and decide if the listing should be deleted or they should receive a personal follow-up. Notably, there is no automatic deletion; it's always a human making that decision. The number of contractors and services above is not too big as to make something like that infeasible.

I think this approach sounds great. 

> Send 3 nag emails in some periodic time frame (like 1 week)
> On the 4th check, if no change has been made, take some action like unpublish the node or block the user or uncheck the checkbox
> After the above action(s) is(are) taken, send a notice (maybe only weekly?) to an admin or admin group about the latest actions taken

I love this -- as long as nobody installs the rules module on b.org :D

> I believe that it is possible to add simple checkbox on a user profile

I don't think a field is the right solution. I have another idea. What if rather than modifying the user profile or the organization node directly, they got added to a queue of items to review? 

We can use the "Last access" date to determine if someone is active, we don't need to ask them to do anything other than review the setting (and maybe put a link in the email that takes them right to it via `?destination` to make things easy).

Organization profiles do a lot of different things though, and a group that provides hosting in addition to Backdrop development might still want to appear on the hosting page even if they are not taking on development projects. Unpublishing those nodes because one of the options is wrong seems like overkill. And if all the options are correct, then we can't use the "last modified" date, as someone might review their org page but not make any changes... What if we added them to a queue of orgs to review manually?  Oooh! I like this. We could do that for

**Comment by irinaz** [comment]
@jenlampton , this sounds like a great plan!  What would be timeframe to implement it?  Who can work on this or help with this?  If there is a marketing effort for Backdrop and people begin looking at service providers it critical that listings are accurate.  Also, can we add an option to add contact info for service providers?

**Comment by yorkshire-pudding** [comment]
@jenlampton - this idea sounds good.  How about a field on the org node that is "Last reviewed" and perhaps we automatically populate it somehow as part of this workflow?  This is quite common for websites to show and confirm that the information has been reviewed and confirmed to be correct.

**Comment by albanycomputers** [comment]
I'm happy to help.

**Comment by jenlampton** [comment]
> this idea sounds good. How about a field on the org node that is "Last reviewed" and perhaps we automatically populate it somehow as part of this workflow? This is quite common for websites to show and confirm that the information has been reviewed and confirmed to be correct.

We could set the date for all orgs that offer development services, on `node_save()` (and via update hook for all the current ones).   Then we could build it into the "removed from queue" to update it.

**Comment by albanycomputers** [comment]
@jenlampton ... Sorry, only just seen this...  I'll have a read...

**Comment by albanycomputers** [comment]
If it's going to be a contrib module, then I feel it should be more useful than just for B.org... and if possible be more useful to the wider community.

Plus...  I must be missing something... whats borg_mailers module? :)

**Comment by yorkshire-pudding** [comment]
> Plus... I must be missing something... whats borg_mailers module? :)

https://github.com/backdrop-ops/backdropcms.org/tree/main/www/modules/custom/borg_mailers

**Comment by irinaz** [comment]
@albanycomputers , this task is very specific to b.org , it might be more efficient to consider it a custom module rather than a full contrib module.

**Comment by albanycomputers** [comment]
@yorkshire-pudding Thanks... I did search for it... but obviously not good enough :)

**Comment by stpaultim** [comment]
I think that something like this would be pretty helpful. 

I agree, that the needs might be too specific to make a custom module, unless someone is really motivated. Would be quicker to do something custom. 

Do we have any volunteers?

**Comment by albanycomputers** [comment]
Is there a b.org version I can play with to become familiar with the fields available etc...

**Comment by irinaz** [comment]
@albanycomputers ,  is this the same module that you are asking https://github.com/backdrop-ops/backdropcms.org/tree/main/www/modules/custom/borg_mailers or you are looking for something else?

**Comment by albanycomputers** [comment]
@irinaz Yes, It's the module that Jen was speaking about above... I'd not investigated that far into Borg territory before and I couldn't find it on the repos...  

Found it now thanks.

**Comment by bugfolder** [comment]
Hi @albanycomputers, if you're going to be working on a module that is intended for backdropcms.org, I'd recommend that you set up a local installation of a copy of the site for your development, which you can do using sanitized versions of the backdrop and civicrm databases and files. The codebase (which I see you've found) contains `borg_mailers` module as well as several other b.org-specific custom modules. I can help you with getting it set up; if you want to go this route, contact me via DM on Zulip.

**Comment by irinaz** [comment]
What is best way to organized the team to work on this?  @bugfolder , @yorkshire-pudding @albanycomputers ?

**Comment by albanycomputers** [comment]
@bugfolder Many thanks...  I'll have a look tomorrow, where can I get a sanitized version?

**Comment by bugfolder** [comment]
@albanycomputers, link sent via Zulip DM. There's probably going to be some back-and-forth to get things configured, so let's continue the nuts-and-bolts discussion there.

**Comment by albanycomputers** [comment]
Tried to setup on a subdomain but failed due to permissions on importing the civi db... so will setup a local server over the weekend and give it another try.

**Comment by albanycomputers** [comment]
Tried with a local copy and had it working to a degree...  the theme was not working and when I tried to access a link I received a database connection error...

I don't have the time available to try and get a local copy working.  I've spent just over 10 hours on this already and haven't gotten anywhere :(

Is there a different way I can help?

**Comment by jenlampton** [comment]
@albanycomputers I suspect the problem is civi, which is a lot more particular than Backdrop to install. Depending on what you are working on, you can `UPDATE system SET status = 0 WHERE name = 'civicrm';` and get backdrop running without civi. (This works best for developing modules that don't touch user accounts.)

**Comment by albanycomputers** [comment]
@jenlampton yeah, I suspect the same... and I could run it from a subdomain... would make it a lot easier for me to play with :)

**Comment by irinaz** [comment]
Looks like this effort is stuck because of complexity of the site.  @jenlampton , do you think it might be worth to simplify site setup?

**Comment by jenlampton** [comment]
I am a huge fan of simplicity, but saying "simplify site setup" isn't really helpful when addressing a specific concern. In the case of b.org, everything that's there exists for a reason, nothing has been added "just because". If we are going to remove thing X, then we need an alternative solution for the need that caused thing X to be added.

**Comment by albanycomputers** [comment]
I am planning on having another bash at this, but the holiday and getting back into the swing of things has pushed me back a bit.

**Comment by bugfolder** [comment]
I'd be happy to jump on a Zoom, Zulip, or Jitsi with you, @albanycomputers, to help get your local site set up (timezones and schedules permitting). I realize the initial setup can be fiddly (what with CiviCRM and two dbs), but once it's initially configured, doing and up-to-date mirror is doable with a relatively simple script.

**Comment by irinaz** [comment]
> I am a huge fan of simplicity, but saying "simplify site setup" isn't really helpful when addressing a specific concern. In the case of b.org, everything that's there exists for a reason, nothing has been added "just because". If we are going to remove thing X, then we need an alternative solution for the need that caused thing X to be added.

Seems that CiviCRM creates some complexity with local setup - is it used on b.org or on events. backdropcms.org?

**Comment by bugfolder** [comment]
> is it used on b.org or on events.backdropcms.org?

Only on b.org. It is currently used to handle various bulk mailings (keeping track of who gets them, allowing users to opt-in or opt-out). Civi has the capability to handle event registrations (including taking payments), so that could conceivably be in the future (though there would be some issues to work out vis-a-vis events.b.org being a separate installation).

**Comment by irinaz** [comment]
@bugfolder , we spend quite some time together trying to complete local setup on my computer.  Do you think that it would be possible to replace Civi on the main site with https://backdropcms.org/project/simplenews to handle bulk mailings ?

**Comment by jenlampton** [comment]
I do not think we should be discussing removing civicrm from the site.
Especially not in this issue. Let's work towards making progress, rather
than moving backwards.

**Comment by irinaz** [comment]
@jenlampton , @yorkshire-pudding , what would be next step to do validate that records https://backdropcms.org/support/services/contractors?expertise=All&country=All and https://backdropcms.org/support/services ?  It looks like there is more demand for BackdropCMS in 2024.  
I can manually check status of all websites listed on the page, maybe admin can unpublish sites that are not responding, like https://sentaidigital.com/  or are not really relevant to Backdrop https://linuxstans.com/?s=backdrop ?

**Comment by yorkshire-pudding** [comment]
@irinaz - I think this is with @albanycomputers on the custom module to support the annual review process.

I wonder though whether, perhaps, in the absence of a custom module, we could do a manual check and sense check and remove any obviously spam/expired records. What do you think @jenlampton @bugfolder @stpaultim ?  Perhaps an admin view could be made so volunteers like @irinaz and I could flag ones and add a reason and then someone with user admin permission can deselect that "Available for Hire" box? Or could be even simpler and a privately shared google sheet?

**Comment by albanycomputers** [comment]
It's still on my to-do list, but as always... time is the problem... January is no easier... I have my company accounts to finalise before the end of Jan :-(

**Comment by jenlampton** [comment]
If you can identify people who might not be active, we can contact them to
follow-up. If we don't get a reply we can unpublished them.

Maybe let's create another issue for the manual review?
